src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/HalfConverter.cs

```diff
@@ -0,0 +1,228 @@
+
+                    return value;
+                }
+            }
+
+            return Read(ref reader, Type, options);
+        }
+
+        internal override void WriteNumberWithCustomHandling(Utf8JsonWriter writer, Half value, JsonNumberHandling handling)
+        {
+            if ((JsonNumberHandling.WriteAsString & handling) != 0)
+            {
+#if NET8_0_OR_GREATER
+                const byte Quote = JsonConstants.Quote;
+                Span<byte> buffer = stackalloc byte[MaxFormatLength + 2];
+#else
+                const char Quote = (char)JsonConstants.Quote;
+                Span<char> buffer = stackalloc char[MaxFormatLength + 2];
+#endif
+                buffer[0] = Quote;
+                Format(buffer.Slice(1), value, out int written);
+
+                int length = written + 2;
+                buffer[length - 1] = Quote;
+                writer.WriteRawValue(buffer.Slice(0, length));
+            }
+            else if ((JsonNumberHandling.AllowNamedFloatingPointLiterals & handling) != 0)
+            {
+                WriteFloatingPointConstant(writer, value);
+            }
+            else
+            {
+                WriteCore(writer, value);
+            }
+        }
+
+        private static bool TryGetFloatingPointConstant(ref Utf8JsonReader reader, out Half value)
+        {
+            Span<byte> buffer = stackalloc byte[MaxFormatLength];
+            int written = reader.CopyValue(buffer);
+
+            return JsonReaderHelper.TryGetFloatingPointConstant(buffer.Slice(0, written), out value);
+        }
+
+        private static void WriteFloatingPointConstant(Utf8JsonWriter writer, Half value)
+        {
+            if (Half.IsNaN(value))
+            {
+                writer.WriteNumberValueAsStringUnescaped(JsonConstants.NaNValue);
+            }
+            else if (Half.IsPositiveInfinity(value))
+            {
+                writer.WriteNumberValueAsStringUnescaped(JsonConstants.PositiveInfinityValue);
+            }
+            else if (Half.IsNegativeInfinity(value))
+            {
+                writer.WriteNumberValueAsStringUnescaped(JsonConstants.NegativeInfinityValue);
+            }
+            else
+            {
+                WriteCore(writer, value);
+            }
+        }
+
+        // Half.TryFormat/TryParse(ROS<byte>) are not available on .NET 7
+        // we need to use Half.TryFormat/TryParse(ROS<char>) in that case.
+        private static bool TryParse(
+#if NET8_0_OR_GREATER
+            ReadOnlySpan<byte> buffer,
+#else
+            ReadOnlySpan<char> buffer,
+#endif
+            out Half result)
+        {
+            bool success = Half.TryParse(buffer, NumberStyles.Float | NumberStyles.AllowThousands, CultureInfo.InvariantCulture, out result);
+
+            // Half.TryParse is more lax with floating-point literals than other S.T.Json floating-point types
+            // e.g: it parses "naN" successfully. Only succeed with the exact match.
```

Comment:
If I'm understanding this correctly, Trying to deserialize `{ "number" : NaN }` will be successful whereas `{ "number" : naN }` will fail.

However both cases are invalid JSON since non-numeric identifiers can only be contained in JSON strings.

Given that this method will get called in cases where you have `JsonTokenType.Number`, it should always reject non-numeric identifiers.

Reply:
> Trying to deserialize { "number" : NaN } will be successful whereas { "number" : naN } will fail.

No, they need be inside quotes, otherwise the reader would throw invalid token:

```
var reader = new Utf8JsonReader(Encoding.UTF8.GetBytes("NaN"));
Console.WriteLine(reader.Read());
// throws: Unhandled exception. System.Text.Json.JsonReaderException: 'N' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
```

Reply:
> Given that this method will get called in cases where you have JsonTokenType.Number

It can actually be called when TokenType is JsonTokenType.String or JsonTokenType.PropertyName too when `ReadCore` is called from `ReadAsPropertyNameCore` or `ReadNumberWithCustomHandling`.